File: tests/uart/uart.py
"""
source: https://lab.whitequark.org/notes/2016-10-18/implementing-an-uart-in-verilog-and-migen/

instance: UART(clk_freq=4800, baud_rate=1200)

Removed _TestPads() Module, in UART() replaced serial variable with rx, tx signals 
    directly as inputs

"""


from migen import *
from migen.genlib.fsm import *


def _divisor(freq_in, freq_out, max_ppm=None):
    divisor = freq_in // freq_out
    if divisor <= 0:
        raise ArgumentError("output frequency is too high")

    ppm = 1000000 * ((freq_in / divisor) - freq_out) / freq_out
    if max_ppm is not None and ppm > max_ppm:
        raise ArgumentError("output frequency deviation is too high")

    return divisor


class UART(Module):
    def __init__(self, clk_freq, baud_rate):
        self.rx_data = Signal(8)
        self.rx_ready = Signal()
        self.rx_ack = Signal()
        self.rx_error = Signal()

        self.tx_data = Signal(8)
        self.tx_ready = Signal()
        self.tx_ack = Signal()

        self.rx = Signal(reset=1)
        self.tx = Signal()


        divisor = _divisor(freq_in=clk_freq, freq_out=baud_rate, max_ppm=50000)

        ###

        rx_counter = Signal(max=divisor)
        self.rx_strobe = rx_strobe = Signal()
        self.comb += rx_strobe.eq(rx_counter == 0)
        self.sync += \
            If(rx_counter == 0,
                rx_counter.eq(divisor - 1)
            ).Else(
                rx_counter.eq(rx_counter - 1)
            )

        self.rx_bitno = rx_bitno = Signal(3)
        self.submodules.rx_fsm = FSM(reset_state="IDLE")
        self.rx_fsm.act("IDLE",
            If(~self.rx,
                NextValue(rx_counter, divisor // 2),
                NextState("START")
            )
        )
        self.rx_fsm.act("START",
            If(rx_strobe,
                NextState("DATA")
            )
        )
        self.rx_fsm.act("DATA",
            If(rx_strobe,
                NextValue(self.rx_data, Cat(self.rx_data[1:8], self.rx)),
                NextValue(rx_bitno, rx_bitno + 1),
                If(rx_bitno == 7,
                    NextState("STOP")
                )
            )
        )
        self.rx_fsm.act("STOP",
            If(rx_strobe,
                If(~self.rx,
                    NextState("ERROR")
                ).Else(
                    NextState("FULL")
                )
            )
        )
        self.rx_fsm.act("FULL",
            self.rx_ready.eq(1),
            If(self.rx_ack,
                NextState("IDLE")
            ).Elif(~self.rx,
                NextState("ERROR")
            )
        )
        self.rx_fsm.act("ERROR",
            self.rx_error.eq(1))

        ###

        tx_counter = Signal(max=divisor)
        self.tx_strobe = tx_strobe = Signal()
        self.comb += tx_strobe.eq(tx_counter == 0)
        self.sync += \
            If(tx_counter == 0,
                tx_counter.eq(divisor - 1)
            ).Else(
                tx_counter.eq(tx_counter - 1)
            )

        self.tx_bitno = tx_bitno = Signal(3)
        self.tx_latch = tx_latch = Signal(8)
        self.submodules.tx_fsm = FSM(reset_state="IDLE")
        self.tx_fsm.act("IDLE",
            self.tx_ack.eq(1),
            If(self.tx_ready,
                NextValue(tx_counter, divisor - 1),
                NextValue(tx_latch, self.tx_data),
                NextState("START")
            ).Else(
                NextValue(self.tx, 1)
            )
        )
        self.tx_fsm.act("START",
            If(self.tx_strobe,
                NextValue(self.tx, 0),
                NextState("DATA")
            )
        )
        self.tx_fsm.act("DATA",
            If(self.tx_strobe,
                NextValue(self.tx, tx_latch[0]),
                NextValue(tx_latch, Cat(tx_latch[1:8], 0)),
                NextValue(tx_bitno, tx_bitno + 1),
                If(self.tx_bitno == 7,
                    NextState("STOP")
                )
            )
        )
        self.tx_fsm.act("STOP",
            If(self.tx_strobe,
                NextValue(self.tx, 1),
                NextState("IDLE")
            )
        )

File: tests/uart/uart.rb
require 'ruby_rtl'
 
include RubyRTL
 
class Uart < Circuit
  def initialize
    input :sys_rst => :bv1
    input :sys_clk => :bv1
    wire :[:name, :rx_data] => :
    wire :[:name, :rx_ready] => :
    wire :[:name, :rx_ack] => :
    wire :[:name, :next_value] => :
    wire :[:name, :rx_error] => :
    wire :[:name, :next_value_ce] => :
    wire :[:name, :tx_data] => :
    wire :[:name, :tx_ready] => :
    wire :[:name, :tx_ack] => :
    wire :[:name, :rx] => :
    wire :[:name, :tx] => :
    wire :[:name, :rx_counter] => :
    wire :[:name, :rx_strobe] => :
    wire :[:name, :rst] => :
    wire :[:name, :rx_bitno] => :
    wire :[:name, :tx_counter] => :
    wire :[:name, :tx_strobe] => :
    wire :[:name, :tx_bitno] => :
    wire :[:name, :tx_latch] => :
    wire :[:name, :state] => :
    wire :[:name, :next_state] => :
 
    assign(rx_strobe <= (rx_counter == 0))
    assign(tx_strobe <= (tx_counter == 0))
    combinatorial(){
      assign(rx_error <= 0)
      assign(fsm0_next_state <= 0)
      assign(rx_counter_next_value0 <= 0)
      assign(rx_counter_next_value_ce0 <= 0)
      assign(rx_data_next_value1 <= 0)
      assign(rx_data_next_value_ce1 <= 0)
      assign(rx_bitno_next_value2 <= 0)
      assign(rx_bitno_next_value_ce2 <= 0)
      assign(fsm0_next_state <= fsm0_state)
      Case(fsm0_state){
        When(1){
          If((rx_strobe == 1)){
            assign(fsm0_next_state <= 2)
          }
        }
        When(2){
          If((rx_strobe == 1)){
            assign(rx_data_next_value1 <= (rx_data[7..1] concat rx))
            assign(rx_data_next_value_ce1 <= 1)
            assign(rx_bitno_next_value2 <= (rx_bitno add 1))
            assign(rx_bitno_next_value_ce2 <= 1)
            If((rx_bitno == 7)){
              assign(fsm0_next_state <= 3)
            }
          }
        }
        When(3){
          If((rx_strobe == 1)){
            If((rx != 1)){
              assign(fsm0_next_state <= 5)
            Else{
              assign(fsm0_next_state <= 4)
            }
          }
        }
        When(5){
          assign(rx_error <= 1)
        }
        When(4){
          assign(rx_ready <= 1)
          If((rx_ack == 1)){
            assign(fsm0_next_state <= 0)
          Else{
            If((rx != 1)){
              assign(fsm0_next_state <= 5)
            }
          }
        }
        Else{
          If((rx != 1)){
            assign(rx_counter_next_value0 <= 2)
            assign(rx_counter_next_value_ce0 <= 1)
            assign(fsm0_next_state <= 1)
          }
        }
      }
    }
    combinatorial(){
      assign(fsm1_next_state <= 0)
      assign(tx_counter_t_next_value0 <= 0)
      assign(tx_counter_t_next_value_ce0 <= 0)
      assign(tx_latch_t_next_value1 <= 0)
      assign(tx_latch_t_next_value_ce1 <= 0)
      assign(tx_f_next_value <= 0)
      assign(tx_f_next_value_ce <= 0)
      assign(tx_bitno_t_next_value2 <= 0)
      assign(tx_bitno_t_next_value_ce2 <= 0)
      assign(fsm1_next_state <= fsm1_state)
      Case(fsm1_state){
        When(1){
          If((tx_strobe == 1)){
            assign(tx_f_next_value <= 0)
            assign(tx_f_next_value_ce <= 1)
            assign(fsm1_next_state <= 2)
          }
        }
        When(2){
          If((tx_strobe == 1)){
            assign(tx_f_next_value <= tx_latch[0..0])
            assign(tx_f_next_value_ce <= 1)
            assign(tx_latch_t_next_value1 <= (tx_latch[7..1] concat 0))
            assign(tx_latch_t_next_value_ce1 <= 1)
            assign(tx_bitno_t_next_value2 <= (tx_bitno add 1))
            assign(tx_bitno_t_next_value_ce2 <= 1)
            If((tx_bitno == 7)){
              assign(fsm1_next_state <= 3)
            }
          }
        }
        When(3){
          If((tx_strobe == 1)){
            assign(tx_f_next_value <= 1)
            assign(tx_f_next_value_ce <= 1)
            assign(fsm1_next_state <= 0)
          }
        }
        Else{
          assign(tx_ack <= 1)
          If((tx_ready == 1)){
            assign(tx_counter_t_next_value0 <= 3)
            assign(tx_counter_t_next_value_ce0 <= 1)
            assign(tx_latch_t_next_value1 <= tx_data)
            assign(tx_latch_t_next_value_ce1 <= 1)
            assign(fsm1_next_state <= 1)
          Else{
            assign(tx_f_next_value <= 1)
            assign(tx_f_next_value_ce <= 1)
          }
        }
      }
    }
    sequential(){
      If((rx_counter == 0)){
        assign(rx_counter <= 3)
      Else{
        assign(rx_counter <= (rx_counter sub 1))
      }
      If((tx_counter == 0)){
        assign(tx_counter <= 3)
      Else{
        assign(tx_counter <= (tx_counter sub 1))
      }
      assign(fsm0_state <= fsm0_next_state)
      If((rx_counter_next_value_ce0 == 1)){
        assign(rx_counter <= rx_counter_next_value0)
      }
      If((rx_data_next_value_ce1 == 1)){
        assign(rx_data <= rx_data_next_value1)
      }
      If((rx_bitno_next_value_ce2 == 1)){
        assign(rx_bitno <= rx_bitno_next_value2)
      }
      assign(fsm1_state <= fsm1_next_state)
      If((tx_counter_t_next_value_ce0 == 1)){
        assign(tx_counter <= tx_counter_t_next_value0)
      }
      If((tx_latch_t_next_value_ce1 == 1)){
        assign(tx_latch <= tx_latch_t_next_value1)
      }
      If((tx_f_next_value_ce == 1)){
        assign(tx <= tx_f_next_value)
      }
      If((tx_bitno_t_next_value_ce2 == 1)){
        assign(tx_bitno <= tx_bitno_t_next_value2)
      }
      If((sys_rst == 1)){
        assign(rx_data <= 0)
        assign(tx <= 0)
        assign(rx_counter <= 0)
        assign(rx_bitno <= 0)
        assign(tx_counter <= 0)
        assign(tx_bitno <= 0)
        assign(tx_latch <= 0)
        assign(fsm0_state <= 0)
        assign(fsm1_state <= 0)
      }
    }
  end
end
